### PR TITLE
fix commit logic not to regenerate workingset

### PR DESCRIPTION
**Pull request by koseoyoung**
works on #1872 

The issue is that even though we failed to validate, we are able to commit the block by regenerating the workingset. So I changed that if workingset doens't exist, not regenerate but return error. 

Impact on this change
1. StartExistingBlockchain() -- add validate before commit 
2. Consensus: if workingset cache has been reset because of offline -- error handling by revalidate/commit

### CODE REVIEW

File: consensus/scheme/rolldpos/rolldposctx.go [hunk 1]
```diff
@@ -484,6 +485,14 @@ func (ctx *rollDPoSCtx) Commit(msg interface{}) (bool, error) {
...
 	case blockchain.ErrInvalidTipHeight:
 		return true, nil
+	case factory.ErrNotValidated:
+		// if the cache has been reset, we need to revalidate/commit the block
+		if err := ctx.chain.ValidateBlock(pendingBlock); err != nil {
+			return false, err
+		}
+		if err := ctx.chain.CommitBlock(pendingBlock); err != nil {
+			return false, err
+		}
```
Comment (CoderZhi) [inline consensus/scheme/rolldpos/rolldposctx.go]: as discussed offline, these code won't be hit.

Comment (koseoyoung) [reply consensus/scheme/rolldpos/rolldposctx.go]: yes! moved error handling logic into blockchain.go commitblock()

**Comment by koseoyoung** [comment]
@dustinxie After discussion with zhi, the unit test seems not correct(bigint_test.go) We shouldn’t commit the block which failed to validate. And Commit() is ok to regenerate workingset when we assume the block is valid. Can we delete that part where it tries to commit again even though validation fails in the unit test?

File: blockchain/blockchain.go [hunk 2]
```diff
@@ -599,7 +602,13 @@ func (bc *blockchain) commitBlock(ctx context.Context, blk *block.Block) error {
 	sfTimer := bc.timerFactory.NewTimer("sf.Commit")
 	err = bc.sf.Commit(ctx, blk)
 	sfTimer.End()
-	// detach working set so it can be freed by GC
+	if errors.Cause(err) == factory.ErrNotValidated {
```
Comment (dustinxie) [inline blockchain/blockchain.go]: this is kine of strange, Commit() --> ErrNotValidated --> then re-commit?
can there be a better way?

File: blockchain/blockchain.go [hunk 3]
```diff
@@ -534,6 +534,9 @@ func (bc *blockchain) startExistingBlockchain(ctx context.Context) error {
 			return err
 		}
 		ctx = bc.contextWithBlock(ctx, producer, blk.Height(), blk.Timestamp())
+		if err := bc.sf.Validate(ctx, blk); err != nil {
```
Comment (dustinxie) [inline blockchain/blockchain.go]: why need to add this?
why the code can work now w/o it?

**Comment by dustinxie** [comment]
> @dustinxie After discussion with zhi, the unit test seems not correct(bigint_test.go) We shouldn’t commit the block which failed to validate. And Commit() is ok to regenerate workingset when we assume the block is valid. Can we delete that part where it tries to commit again even though validation fails in the unit test?

yes, we should understand the test logic, and modify accordingly

basically, it need to be able to detect a negative transfer, and not let such block to be committed

File: blockchain/blockchain.go (hunk 2 above)
Comment (dustinxie) [reply blockchain/blockchain.go]: can we nil the workingset, when validation fail?

factory can regenerate the workingset, but it will fail upon Commit()

**Comment by koseoyoung** [comment]
I tried to have a validationFailFlag with workingset inside of state factory to set up the flag true if a block fails to be validated and remember the result when try to commit again. But it has a problem because negative transfer fails not in factory.Validate() but in blockvalidator side.

**Pull request closed without merging**